test/extended/cli/builds.go

```diff
@@ -0,0 +1,339 @@
+package cli
+
+import (
+	"encoding/json"
+	"fmt"
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+	t "github.com/onsi/gomega/types"
+	exutil "github.com/openshift/origin/test/extended/util"
+	"k8s.io/apimachinery/pkg/util/wait"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+	"regexp"
+	"strings"
+	"time"
+)
+
+const (
+	bcOutputTemplate       = "{{with .spec.output.to}}{{.kind}} {{.name}}{{end}}"
+	bcOutputToNameTemplate = "{{ .spec.output.to.name }}"
+	bcSourceTypeTemplate   = "{{ .spec.source.type }}"
+	testDockerFileContent  = "FROM image-registry.openshift-image-registry.svc:5000/openshift/tests:latest"
+)
+
+func waitForBuild(oc *exutil.CLI, bcName, buildNumber string) error {
+	args := []string{"builds"}
+
+	if buildNumber != "" {
+		args = append(args, fmt.Sprintf("%v-%v", bcName, buildNumber))
+	} else {
+		args = append(args, "-l", fmt.Sprintf("openshift.io/build-config.name=%v", bcName))
+	}
+
+	return wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+		out, err := oc.Run("get").Args(args...).Output()
+
+		if err != nil {
+			return false, nil
+		}
+		lines := strings.Split(out, "\n")
+
+		return len(lines) >= 2, nil
+	})
+}
+
+func getExpectedBCOutputMatcher(bcName string, isNames ...string) t.GomegaMatcher {
+	matcher := o.ContainSubstring("Success")
+	if bcName != "" {
+		matcher = o.And(matcher, o.ContainSubstring(fmt.Sprintf("buildconfig.build.openshift.io \"%v\" created", bcName)))
+	} else {
+		e2e.Logf("no bcName specified")
+	}
+
+	for _, isName := range isNames {
+		if isName != "" {
+			matcher = o.And(matcher, o.ContainSubstring(fmt.Sprintf("imagestream.image.openshift.io \"%v\" created", isName)))
+		}
+	}
+	return matcher
+}
+
+func getBCInfoByTemplate(oc *exutil.CLI, bcName, template string) string {
+	if bcName == "" {
+		e2e.Logf("no bcName specified")
+		return ""
+	}
+	out, err := oc.Run("get").Args("bc/"+bcName, "--template", template).Output()
+
+	if err != nil {
+		e2e.Logf("failed to get %v: %v", bcName, err)
+		return ""
+	}
+	return out
+}
+
+func getBCSourceType(oc *exutil.CLI, bcName string) string {
+	return getBCInfoByTemplate(oc, bcName, bcSourceTypeTemplate)
+}
+
+func getBCOutputType(oc *exutil.CLI, bcName string) string {
+	return getBCInfoByTemplate(oc, bcName, bcOutputTemplate)
+}
+
+var _ = g.Describe("[sig-cli] oc builds", func() {
+	defer g.GinkgoRecover()
+
+	oc := exutil.NewCLI("oc-builds").AsAdmin()
+
+	g.Describe("new-build", func() {
+		g.It("build from a binary with no inputs requires name", func() {
```

Comment:
Yeah, I agree with @ravisantoshgudimetla, I'd change those `It`'s to `By`'s.

Reply:
Also, it's nice and simplifies review if the `By` description matches that from https://github.com/openshift/origin/blob/d704a4d2ab5e55731d11770c11eacd666940b944/test/extended/testdata/cmd/test/cmd/builds.sh#L31 which it does here, but not necessarily everywhere.

Reply:
My suggestion for organization is:
1. `Describe` as the top level
2. `It` as grouping, for example you have `new-build`, `complex-build`, etc.
3. `By` to mark every "group" from the original file